drivers/platform/surface/surface_fan.c

```diff
@@ -0,0 +1,146 @@
+// SPDX-License-Identifier: GPL-2.0+
+/*
+ * Surface Fan driver for Surface System Aggregator Module. It provides access
+ * to the fan's rpm through the hwmon system.
+ *
+ * Copyright (C) 2023 Ivor Wanders <ivor@iwanders.net>
+ */
+
+#include <linux/hwmon.h>
+#include <linux/kernel.h>
+#include <linux/module.h>
+#include <linux/platform_device.h>
+#include <linux/surface_aggregator/device.h>
+#include <linux/types.h>
+
+#define SURFACE_FAN_MIN_SPEED 2000
+#define SURFACE_FAN_MAX_SPEED 7500
+
+struct fan_data {
+	struct device *dev;
+	struct ssam_device *sdev;
+	struct device *hdev;
+};
+
+// SSAM
+SSAM_DEFINE_SYNC_REQUEST_R(__ssam_fan_get, __le16,
+			   {
+				   .target_category = SSAM_SSH_TC_FAN,
+				   .target_id = SSAM_SSH_TID_SAM,
+				   .command_id = 0x01,
+				   .instance_id = 0x01,
+			   });
+
+// hwmon
+umode_t surface_fan_hwmon_is_visible(const void *drvdata,
+				     enum hwmon_sensor_types type, u32 attr,
+				     int channel)
+{
+	switch (type) {
+	case hwmon_fan:
+		switch (attr) {
+		case hwmon_fan_input:
+		case hwmon_fan_label:
+		case hwmon_fan_min:
+		case hwmon_fan_max:
+			return 0444;
+		default:
+			break;
+		}
+	default:
+		break;
+	}
+	return 0;
+}
+
+static int surface_fan_hwmon_read(struct device *dev,
+				  enum hwmon_sensor_types type, u32 attr,
+				  int channel, long *val)
+{
+	struct fan_data *d = dev_get_drvdata(dev);
+	__le16 value;
+	int res;
+
+	switch (type) {
+	case hwmon_fan:
+		switch (attr) {
+		case hwmon_fan_input:
+			res = __ssam_fan_get(d->sdev->ctrl, &value);
+			if (res) {
+				return -EIO;
+			}
+			*val = le16_to_cpu(value);
+			return 0;
+		case hwmon_fan_min:
+			*val = SURFACE_FAN_MIN_SPEED;
+			return 0;
+		case hwmon_fan_max:
+			*val = SURFACE_FAN_MAX_SPEED;
+			return 0;
+		default:
+			break;
+		}
+		break;
+	default:
+		break;
+	}
+
+	return -1;
+}
+
+static const struct hwmon_channel_info *const surface_fan_info[] = {
+	HWMON_CHANNEL_INFO(fan, HWMON_F_INPUT | HWMON_F_MAX | HWMON_F_MIN), NULL
```

Comment:
Probably split this across two lines, with the `NULL` on the second one.

Reply:
Changed with 536311935e4391075b93fff5c2eb88f6cdf95edd